mod_form.php

```diff
@@ -298,87 +310,75 @@ public function definition() {
+            'section' => 'coursecustomfieldmapping',
+            'hasid' => true,
+            'hasmappings' => isset($coursecustomfieldmappings),
+            'accredibleoptions' => $accredibleoptions,
+            'moodleoptions' => $coursecustomfieldoptions,
+            'nocoursecustomoptions' => count($coursecustomfieldoptions) == 1,
+        ];
+
+        $mform->addElement('html', $OUTPUT->render_from_template('mod_accredible/mappings', $coursecustomfieldmappingcontent));
 
         // Attribute mapping: user profile fields.
         $mform->addElement('header',
             'attributemappinguserprofilefields',
             get_string('attributemappinguserprofilefields', 'accredible')
         );
-        $mform->addElement(
-            'select',
-            'userprofilefieldmapping[0][id]',
-            get_string('moodleuserprofilefield', 'accredible'),
-            $formhelper->load_user_profile_field_options(),
-            $inputstyle
-        );
-        $this->set_mapping_field_default($mform, $attributemappingdefaultvalues, 'userprofilefieldmapping', 'id', 0);
 
-        $mform->addElement(
-            'select',
-            'userprofilefieldmapping[0][accredibleattribute]',
-            get_string('accrediblecustomattributename', 'accredible'),
-            $attributekeyschoices,
-            $inputstyle
-        );
-        $this->set_mapping_field_default(
-            $mform,
-            $attributemappingdefaultvalues,
-            'userprofilefieldmapping',
-            'accredibleattribute',
-            0
-        );
+        $userprofilefieldmappings = $attributemappingdefaultvalues['userprofilefieldmapping'];
+        $userprofilefieldmappingcontent = [
+            'mappings' => $userprofilefieldmappings,
+            'section' => 'userprofilefieldmapping',
+            'hasid' => true,
+            'hasmappings' => isset($userprofilefieldmappings),
+            'accredibleoptions' => $accredibleoptions,
+            'moodleoptions' => $userprofilefieldoptions,
+            'noprofileoptions' => count($userprofilefieldoptions) == 1,
+        ];
+
+        $mform->addElement('html', $OUTPUT->render_from_template('mod_accredible/mappings', $userprofilefieldmappingcontent));
 
         $this->standard_coursemodule_elements();
         $this->add_action_buttons();
     }
 
+    /**
+     * Called right before form submission.
+     * We use it to include missing form data from mustache templates.
+     *
+     * @param stdClass $data passed by reference
+     * @return void
+     */
+    public function data_postprocessing($data) {
+        parent::data_postprocessing($data);
+        $submitteddata = $this->_form->getSubmitValues();
+        $data->coursefieldmapping = $submitteddata['coursefieldmapping'];
+        $data->coursecustomfieldmapping = $submitteddata['coursecustomfieldmapping'];
+        $data->userprofilefieldmapping = $submitteddata['userprofilefieldmapping'];
+    }
```

Comment:
This can be addressed later, but I was wondering if we should persist empty entries. 

<img width="925" alt="Screenshot 2024-06-11 at 20 32 47" src="https://github.com/accredible/moodle-mod_accredible/assets/20009662/cd16f7c4-cb3b-4942-b23d-a692ce264aec">